### PR TITLE
Add Other Draw Modes

### PR DESCRIPTION
**Pull request by benaadams**
POINTS/LINES/LINE_LOOP/LINE_STRIP/TRIANGLE_STRIP/TRIANGLE_FAN

Just sanity checking

@deltakosh PTAL does this work?

### CODE REVIEW

File: src/Engine/babylon.engine.ts [hunk 1]
```diff
@@ -5396,4 +5393,16 @@
             }
         }
     }
+
+    export namespace Engine {
```
Comment (deltakosh) [inline src/Engine/babylon.engine.ts]: We are not using enum as they are generating string and string comparison.
Just add statics in the Engine

File: src/Materials/babylon.material.ts [hunk 2]
```diff
@@ -370,6 +402,33 @@
             this.markAsDirty(Material.MiscDirtyFlag);
         }
 
+        public static fillModeToDrawType(fillMode: number): Engine.DrawType {
```
Comment (deltakosh) [inline src/Materials/babylon.material.ts]: Why not using the engine constants directly?

File: src/Mesh/babylon.mesh.ts [hunk 3]
```diff
@@ -997,25 +997,15 @@
...
-                    }
-                    break;
-
-                default:
-                    if (this._unIndexed) {
-                        engine.drawUnIndexed(true, subMesh.verticesStart, subMesh.verticesCount, instancesCount);
-                    } else {
-                        engine.draw(true, subMesh.indexStart, subMesh.indexCount, instancesCount);
-                    }
+            var drawType: Engine.DrawType = Material.fillModeToDrawType(fillMode);
```
Comment (deltakosh) [inline src/Mesh/babylon.mesh.ts]: If using the engine constants we can avoid this cool

File: src/Engine/babylon.engine.ts (hunk 1 above)
Comment (benaadams) [reply src/Engine/babylon.engine.ts]: Wanted to add strong typing rather than just `: number`; but yeah enum does add a lot of trash

Comment (deltakosh) [reply src/Engine/babylon.engine.ts]: Yeah I know :(

**Comment by jbousquie** [comment]
Does this means we will have one day some `Points` public access for the final user (like we have currently for Lines) ?
maybe something like
```javascript
var cloud = BABYLON.MeshBuilder.CreatePointCloud(name, {points: Vector3[], colors?: Color4[]}, scene);
```
then the ability to update/animate them : {instance: cloud}

Would be cool for plenty of needs :-)

**Comment by benaadams** [comment]
Exposes 6 additional modes

Fills for triangles (current)
* TriangleFillMode
* WireFrameFillMode
* PointFillMode

6 additional treatment for vertexes
* PointListDrawMode
* LineListDrawMode
* LineLoopDrawMode
* LineStripDrawMode
* TriangleStripDrawMode
* TriangleFanDrawMode

**Comment by benaadams** [comment]
Had to do some more tweaks, which seem to work - will update PR tomorrow

[image](https://www.youtube.com/watch?v=UAvHrx-kpXU)

**Comment by deltakosh** [comment]
Fantastic!!

File: src/Rendering/babylon.boundingBoxRenderer.ts [hunk 4]
```diff
@@ -85,7 +85,7 @@
                     this._colorShader.bind(worldMatrix);
 
                     // Draw order
-                    engine.draw(false, 0, 24);
+                    engine.drawElementsType(Material.TriangleFillMode, 0, 24);
```
Comment (benaadams) [inline src/Rendering/babylon.boundingBoxRenderer.ts]: oops

File: src/Mesh/babylon.linesMesh.ts [hunk 5]
```diff
@@ -112,7 +112,7 @@
             var engine = this.getScene().getEngine();
 
             // Draw order
-            engine.draw(false, subMesh.indexStart, subMesh.indexCount);
+            engine.drawElementsType(Material.LineListDrawMode, subMesh.indexStart, subMesh.indexCount);
```
Comment (benaadams) [inline src/Mesh/babylon.linesMesh.ts]: oops

**Pull request merged**